packages/pwa-buildpack/lib/WebpackTools/plugins/LocalizationPlugin.js

```diff
@@ -0,0 +1,236 @@
+                    );
+                }
+            }
+        }
+
+        if (!locales) {
+            debug(
+                'No locales found while traversing all modules with i18n flag.'
+            );
+        }
+
+        // Merge all located translation files together and return their paths for a dynamic import
+        const mergedLocalesPaths = await this.writeMergedVirtualLocales(
+            context,
+            locales,
+            inputFileSystem,
+            virtualModules
+        );
+
+        debug('Merged locales into path.', mergedLocalesPaths);
+
+        /**
+         * Build up our importer factory, this provides a global function called fetchLocaleData which in turn completes
+         * a dynamic import of the combined file generated in the step above.
+         * @type {string}
+         */
+        const importerFactory = `function () {
+            return function getLocale(locale) {
+                ${Object.keys(locales)
+                    .map(locale => {
+                        return `if (locale === "${locale}") { 
+                        return import(/* webpackChunkName: "i18n-${locale}" */'${
+                            mergedLocalesPaths[locale]
+                        }');
+                    }`;
+                    })
+                    .join('')}
+                
+                throw new Error('Unable to locate locale ' + locale + ' within generated dist directory.');
+            }
+        }`;
+
+        hooks.afterCompile.tap('LocalizationPlugin', compilation => {
+            // Add global file dependencies for all the found translation files
+            Object.values(locales).forEach(localePaths => {
+                localePaths.forEach(localePath => {
+                    compilation.fileDependencies.add(localePath);
+                });
+            });
+        });
+
+        hooks.emit.tap('LocalizationPlugin', compilation => {
+            // Add individual fileDependencies for each i18n chunk
+            compilation.chunks.forEach(chunk => {
+                if (chunk.name && chunk.name.startsWith('i18n')) {
+                    chunk.getModules().forEach(chunkModule => {
+                        const chunkLocale = path.parse(chunkModule.resource)
+                            .name;
+                        if (
+                            locales[chunkLocale] &&
+                            Array.isArray(locales[chunkLocale])
+                        ) {
+                            locales[chunkLocale].forEach(localePath => {
+                                chunkModule.buildInfo.fileDependencies.add(
+                                    localePath
+                                );
+                            });
+                        }
+                    });
+                }
+            });
+        });
+
+        return `;window.fetchLocaleData = (${importerFactory})()`;
```

Comment:
In my experience, using `window.globalFunction` can lead to issues if any js on the page also happens to redefine this property. Consider namespacing, for example, `window[${storeName}].fetchLocaleData`.

Reply:
I think this just hides the problem, someone can still overwrite `window[${storeName}]` all the same, if someone messes with `fetchLocaleData` they're gonna see the whole app error as it's consumed early on in the `localeProvider`.

Reply:
I guess it's less a problem in the scope of Magento-land where we own the page.

Reply:
I do agree with both of you. How about using `window.__fetchLocaleData__` instead? It is an accepted norm that anything that has a `__` prefix and/or suffix is internal and not messed around with.

Reply:
@revanth0212 good suggestion, I've implemented the change. Maybe we should consider updating the root components version of this at some point, but that'd be a breaking change so maybe not the best idea.